Add DarkMode usage detection support for plugins;

Leonard-The-Wise opened the pull request:
Fix #11546

shriprem commented:
@Leonard-The-Wise, thank you for considering my earlier [suggestion](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11546#issuecomment-1101851045) and implementing `NPPM_GETDARKMODECOLORS`.

I have one more observation on  the usage pattern for `NPPM_GETDARKMODECOLORS`.

Perhaps, we can utilize the WPARAM parameter in the message for a size check. Similar to how in several of Microsoft's Windows API function calls, the `cbSize` parameter is used as a check on the size of a structure whose pointer is also being passed in alongside.

So, the definition for `NPPM_GETDARKMODECOLORS` would look something like:
```C++
	#define NPPM_GETDARKMODECOLORS (NPPMSG + 108)
	// ColorTone NPPM_GETDARKMODECOLORS (size_t cbSize, NppDarkMode::Colors *darkModeColors)
	// cbSize must be set to sizeof(NppDarkMode::Colors)
	// If calling is successful return TRUE, otherwise return FALSE.
```


Then a check can be added in the handling for the `NPPM_GETDARKMODECOLORS` case. Like so:
```C++
		case NPPM_GETDARKMODECOLORS:
		{
			if (static_cast<size_t>(wParam) != sizeof(NppDarkMode::Colors))
			{
				return FALSE;
			}
...
```

Thereby, anyone utilizing this message query in their plugin would hopefully adhere to a usage pattern something like:
```C++
   Theme tCurrent(darkColors);
...
   Colors customColors{};
   SendMessage(nppData._nppHandle, NPPM_GETDARKMODECOLORS, sizeof(customColors), reinterpret_cast<LPARAM>(&customColors));
   tCurrent.change(customColors);
...
```
Thanks again!

shriprem commented:
@Leonard-The-Wise , thank you for implementing it so quickly! 👍

Leonard-The-Wise commented:
I again agree with you to add an extra layer of protection to the main thread by adding this security check to avoid writting on unallocated memory. Yes, Microsoft likes this method for compatibility checks a lot. :)

Implemented.

I also use a method like you posted to change the colors on my plugin. My real code is:

```
        PluginDarkMode::Colors newColors;
        bool bSuccess = Messenger().SendNppMessage<bool>(NPPM_GETDARKMODECOLORS, sizeof(newColors), reinterpret_cast<LPARAM>(&newColors));
        if (bSuccess)
            PluginDarkMode::setThemeColors(newColors);
```

So you guessed right again on that. :)

And you will notice that my `PluginDarkMode` namespace is a ripped-off version of the namespace inside `NppDarkMode.cpp`, but without all the extra includes and dependencies that Notepad++ adds there. I'm actually rewritting that file to facilitate for other people trying to integrate Dark Mode on their own plugins. 

I already "rewrote" the `ListView`, `Toolbar`, `Button` and `Window` subclasses so they will be much easier to integrate into any plugin without any necessity of implementing custom message processings for `WM_CTLCOLOREDIT`, `WM_ERASEBKGND`, `WM_PRINTCLIENT`, `WM_NOTIFY`, etc. 

All the developer will have to do in order to fully implement a Dark Mode inside a Dialog of their own is a simple call to:

```
    PluginDarkMode::autoSetupWindowAndChildren(hWndDialog);
```
I also managed to fix the colors for Push buttons already, so they won't only get black buttons when switching to dark themes.

This, coupled with other functions I implemented to setup high DPI support, they will also be able to call:

```
   _dpiManager.DPIResizeControl(_hSelf);
   _dpiManager.DPIResizeChildren(_hSelf, true);
```

And their windowses and controls will magically scale with DPI settings, adjusting font sizes, etc. Images and icons will be able to scale with DPI also, if they use multiresolution icons and call a `_dpiManager.ScaleIconSize(16)` (being 16 the size in logical pixels) to load the correct image.

Lastly, I also implemented a version of anchoring and docking support for windowses that I also plan to release to raw Win32 C++ plugin developers. All derived from the works on my current Lexer and Compiler plugin.

This is just a sneak peak of what I'm planning to release here. It is outside the scope of this pull request. When the code is more mature I'll offer those classes to @donho for placing it inside the [NPP Plugin Template v4.2](https://github.com/npp-plugins/plugintemplate/releases) for other users to adopt on their own projects.

Meanwhile, let's see what else this pull request should contain.

Cheers.

shriprem commented:
Impressive. Keep up the good work!

I hope that your PR here will get accepted and merged soon into NPP master. It will benefit quite a few plugin authors who have already implemented dark mode support but have done so by having to parse through the settings saved in NPP's config.xml file. With that approach, any changes to the dark mode choices will be reflected in those plugins only after closing and restarting Notepad++.  This PR will address that lag.

Leonard-The-Wise commented on PowerEditor/src/NppDarkMode.cpp:
```diff
@@ -1175,7 +1175,7 @@ namespace NppDarkMode
 
 			LONG_PTR dwStyle = GetWindowLongPtr(hWnd, GWL_STYLE);
-			if ((dwStyle & TCS_BOTTOM) || (dwStyle & TCS_BUTTONS) || (dwStyle & TCS_VERTICAL))
+			if ((dwStyle & TCS_BUTTONS) || (dwStyle & TCS_VERTICAL))
```
This is indeed curious. I tested the tabs with both 32 and 64 bits version of Notepad++ and it contradicts the statement that `This style is not supported if you use ComCtl32.dll version 6`. But I realized that I also use a Windows 64 image and perhaps I'll need to test it in a Windows 32bit VM to confirm.

So, while I'm not sure if the documentation is wrong, I reverted this change.

Leonard-The-Wise replied:
Wait, thinking better... I am indeed using ComCtl32.dll version 6 and my tab do have TCS_BOTTOM and it works. Code on my Control resource file:

```
IDD_LOGGER DIALOGEX 0, 0, 509, 173
STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | WS_CHILD | WS_VISIBLE | WS_CAPTION
CAPTION "NWScript Tools Compiler Console"
FONT 8, "MS Shell Dlg", 400, 0, 0x1
BEGIN
    CONTROL         "",IDC_TABLOGGER,"SysTabControl32",TCS_BOTTOM,2,2,504,167
END
```

ComCtl32.dll version 6 is from Windows Server 2003, nobody today uses any version lower, so it shouldn't matter what windows image I am using, right?

That wouldn't be the first time I see deprecated or incorrect documentation at Microsoft website. So I'm reopening this comment, if you want, I will patch that change again.

donho replied:
There's no harm to keep it, so let it remains as it is.
Thank you for your PR.

Leonard-The-Wise replied:
Ok then, remade it. 👍

Leonard-The-Wise commented on PowerEditor/src/NppDarkMode.cpp:
```diff
@@ -1477,24 +1477,39 @@ namespace NppDarkMode
+								SetWindowTheme(hwnd, p.themeClassName, nullptr);
+							}
+							break;
+						}
```
Added example on issue #11546.
